src/carma-platform/guidance_scripts.py

```diff
@@ -1283,6 +1283,352 @@ def get_planner_trajectory_intervals(
+    if save_stats_dir:
+        stats_dir = Path(save_stats_dir)
+        stats_dir.mkdir(parents=True, exist_ok=True)
+        stats_path = stats_dir / "speed_change_analysis.json"
+        with open(stats_path, "w") as f:
+            json.dump({
+                "pass_results": pass_results,
+                "statistics": statistics
+            }, f, indent=2)
+        print(f"Stats saved to: {stats_path}")
+
+    if save_data_dir:
+        data_dir = Path(save_data_dir)
+        data_dir.mkdir(parents=True, exist_ok=True)
+        data_path = data_dir / "speed_change_analysis_data.npz"
+        np.savez(
+            data_path,
+            timestamps=timestamps,
+            velocities=long_velocities,
+            speed_limits=speed_limits,
+            cmd_velocities=cmd_velocities,
+            speed_limit_changes=speed_limit_changes,
+            response_times=response_times,
+            steady_state_periods=steady_state_periods
+        )
+        print(f"Data saved to: {data_path}")
+
+    if save_plot_dir:
+        plot_dir = Path(save_plot_dir)
+        plot_dir.mkdir(parents=True, exist_ok=True)
+        plot_path = plot_dir / "speed_change_analysis.png"
+        plt.savefig(plot_path, dpi=300)
+        print(f"Plot saved to: {plot_path}")
+    else:
+        plt.show()
+
+    return (pass_results["overall"], statistics, fig, speed_limit_changes, response_times)
+
+
+def detect_speed_limit_changes(timestamps, speed_limits, min_change=0.5):
+    """
+    Detect significant changes in speed limits.
+
+    Args:
+        timestamps: Array of timestamps
+        speed_limits: Array of speed limits
+        min_change: Minimum change in speed limit to be considered significant (m/s)
+
+    Returns:
+        List of tuples (timestamp, old_limit, new_limit) for each detected change
+    """
+    changes = []
+    prev_limit = speed_limits[0]
+
+    for i in range(1, len(speed_limits)):
+        current_limit = speed_limits[i]
+        if prev_limit < 0.01:
+            prev_limit = current_limit
+            continue
+        if abs(current_limit - prev_limit) >= min_change:
+            changes.append((timestamps[i], prev_limit, current_limit))
+            prev_limit = current_limit
+
+    return changes
+
+
+def analyze_speed_responses(timestamps, velocities, speed_limit_changes,
+                           steady_state_duration, speed_tolerance_pct):
+    """
+    Analyze vehicle's response to speed limit changes.
+    Args:
+        timestamps: Array of timestamps
+        velocities: Array of vehicle velocities
+        speed_limit_changes: List of detected speed limit changes
+        steady_state_duration: Duration required at new speed for steady state
+        speed_tolerance_pct: Tolerance percentage for speed matching (used only for steady state)
+    Returns:
+        Tuple containing:
+        - Array of response times for each speed change
+        - List of steady state periods [(start_time, end_time, speed), ...]
```

Comment:
Technically we are only returning the "first X seconds (based on steady_state_duration) of steady state", rather than a list of full steady state periods

Reply:
reworded